testonly/hammer/hammer.go

```diff
@@ -181,15 +185,58 @@ func HitMap(cfg MapConfig) error {
 		}
 	}(ticker.C)
 
-	for count := uint64(1); count < cfg.Operations; count++ {
-		if err := s.retryOneOp(ctx); err != nil {
-			return err
+	var wg sync.WaitGroup
+	errs := make(chan error, cfg.NumCheckers)
+
+	cctx, cancel := context.WithCancel(ctx)
+	for i := 0; i < cfg.NumCheckers; i++ {
+		wg.Add(1)
+		go func(i int) {
+			glog.Infof("%d: start checker %d", s.cfg.MapID, i)
+			defer wg.Done()
+			// Note that we deliberately don't use the cancelable context
+			// cctx so that cancellation in mid-request doesn't give internal
+			// warnings.
+			err := s.readChecker(ctx, cctx.Done())
+			if err != nil {
+				errs <- err
+			}
+			glog.Infof("%d: checker %d done with %v", s.cfg.MapID, i, err)
+		}(i)
+	}
+
+	wg.Add(1)
+	done := make(chan bool)
+	go func() {
+		defer wg.Done()
+		for count := uint64(1); count < cfg.Operations; count++ {
+			err := s.retryOneOp(ctx)
+			if err != nil {
+				errs <- err
+				return
+			}
 		}
+		glog.Infof("%d: completed %d operations on map", cfg.MapID, cfg.Operations)
+		done <- true
+	}()
+
+	var firstErr error
+	select {
+	case <-ctx.Done():
+		glog.Infof("%d: context canceled", cfg.MapID)
+	case <-done:
+		glog.Infof("%d: main hammer goroutine complete", cfg.MapID)
+	case e := <-errs:
+		firstErr = e
+		glog.Infof("%d: first error encountered: %v", cfg.MapID, e)
 	}
-	glog.Infof("%d: completed %d operations on map", cfg.MapID, cfg.Operations)
-	ticker.Stop()
 
-	return nil
+	ticker.Stop()
```

Comment:
Consider consuming all the errors in `errs`

Reply:
This might also make the wait group unneeded.

Reply:
I've added a loop over the `errs`, but I think the wait group is still needed (to be sure that all the checkers have finished and emitted any pending errors).

Reply:
If the number of outputs on `err` was predictable, one could just wait for all of them to appear.

Reply:
I prefer not to have dual-purpose control structures; separating out:
 - the wait group tracks pending goroutines
 - the `errs` channel accumulates error results
makes it easier to figure out what's going on.

Reply:
That separation makes sense.